crates/bevy_ecs/src/archetype.rs

```diff
@@ -15,21 +16,36 @@ use std::{
 };
 
 #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
-pub struct ArchetypeId(usize);
+pub struct ArchetypeId(NonMaxUsize);
 
 impl ArchetypeId {
-    pub const EMPTY: ArchetypeId = ArchetypeId(0);
-    pub const RESOURCE: ArchetypeId = ArchetypeId(1);
-    pub const INVALID: ArchetypeId = ArchetypeId(usize::MAX);
+    pub const EMPTY: ArchetypeId = unsafe { ArchetypeId::new_unchecked(0) };
+    pub const RESOURCE: ArchetypeId = unsafe { ArchetypeId::new_unchecked(1) };
+    pub const INVALID: ArchetypeId = unsafe { ArchetypeId::new_unchecked(usize::MAX - 1) };
 
+    /// Creates a new [`ArchetypeId`].
+    ///
+    /// # Panics
+    /// This will panic if `index` is `usize::MAX`.
     #[inline]
-    pub const fn new(index: usize) -> Self {
-        ArchetypeId(index)
+    pub fn new(index: usize) -> Self {
+        Self(NonMaxUsize::new(index).expect("ArchetypeID cannot be created from usize::MAX"))
     }
 
+    /// Creates a new [`ArchetypeId`] without checking for validity of the
+    /// provided index.
+    ///
+    /// # Safety
+    /// The produced value is only safe if `index` is not `usize::MAX`.
```

Comment:
This safety isn't quite right imo; it's insta-UB to even construct the value with `usize::MAX`.

That is, the safety doc is just "`index` is not `usize::MAX`"

Reply:
```suggestion
    /// `index` must not be `usize::MAX`
```